### PR TITLE
Hierarchical date facet

### LINKED ISSUE #1136: Add hierarchy to the year facet
The blacklight_range_limit slider has accessibility issues. We're looking at implementing a hierarchical facet, similar to vt-arclight:

Screenshot 2024-08-08 at 09 45 59

To get started, see [the implementation there](https://github.com/sul-dlss/vt-arclight/blob/8206bfb9f2fb1c4b52a8682d6efbd97d0a579b0d/app/controllers/catalog_controller.rb#L136-L140).

### PR DESCRIPTION
**Pull request by jmartin-sul**
closes #1136

needs https://github.com/sul-dlss/searchworks_traject_indexer/pull/1521 to populate the field for the facet.

Screenshot 2024-09-05 at 11 45 09 PM

the alignment seems a little weird on the facet removal button (`x`) in that screenshot from my local instance.  but this seems like an improvement over the current facet, so i wonder if we could merge this and refine the CSS if that's the only issue, and the fix isn't totally obvious?

if removing the blacklight_range_limit gem isn't desired afterall, i left that as a separate commit, so easy enough to revert that part or split it out into another PR.

### CODE REVIEW

File: spec/features/blacklight_customizations/hierarchical_date_facet_spec.rb
Comment (marlo-longley) [inline spec/features/blacklight_customizations/hierarchical_date_facet_spec.rb:10]: In a recent PR of mine (https://github.com/sul-dlss/earthworks/pull/1337) we decided that hardcoding fixture counts into the tests was going to cause annoyance in the future.
This is more targeted because it's counting a field that will probably not be updated frequently. But if there was a way to test this without using counts, I'd go for that. If not, I think it's fine.

File: app/components/blacklight/date_facet_hierarchy_component.rb
Comment (marlo-longley) [inline app/components/blacklight/date_facet_hierarchy_component.rb:18]: I have read through this method and understand what you're doing but it is the trickiest part of the code. It might benefit from a comment about what format input is expected, and also about the different cases you handle along the way.

File: app/components/blacklight/date_facet_hierarchy_component.rb
Comment (marlo-longley) [inline app/components/blacklight/date_facet_hierarchy_component.rb:28]: I had a little trouble following when this case would ever be true, because 1) you explicitly set ` tree[century_str].items` to an empty array on line 24, and 2) you've already split the string on `:` on line 22. 

I concluded that this `decade_facet` could only exist if you'd already created it during another loop.

File: spec/features/blacklight_customizations/hierarchical_date_facet_spec.rb
Comment (jmartin-sul) [reply spec/features/blacklight_customizations/hierarchical_date_facet_spec.rb:10]: ah yeah, good point, will think for a bit about trying to get out of using counts...

File: app/components/blacklight/date_facet_hierarchy_component.rb
Comment (jmartin-sul) [reply app/components/blacklight/date_facet_hierarchy_component.rb:18]: good call, will do

File: app/components/blacklight/date_facet_hierarchy_component.rb
Comment (jmartin-sul) [reply app/components/blacklight/date_facet_hierarchy_component.rb:28]: yeah, that's right, this code is looking to update an existing instance if it finds one, or to add that decade for the first time if need be.  will add a comment...

**Review by thatbudakguy — APPROVED** [review]
Tested on stage and ready to merge, pending response to feedback

File: spec/features/blacklight_customizations/hierarchical_date_facet_spec.rb
Comment (jmartin-sul) [reply spec/features/blacklight_customizations/hierarchical_date_facet_spec.rb]: ok, so i wanted to keep somewhat precise search behavior expectations, to confirm that faceting wasn't happening when it shouldn't but also that faceting narrowed results appropriately when applied.  but the worry about test fragility and annoyance totally makes sense.

so given that the main worry in the other PR's discussion seemed to be about frequent annoyance when adding unrelated fixtures, and since it seemed less common that we'd regularly add or remove fixtures in the 1990-1999 date range, i tried to find a middle ground: i kept the hardcoded count for the post-facet results; and for the prefacet expectations, i just looked for the initial total result count, and made sure that stayed unchanged when expanding (but not choosing) a facet in the tree.

hopefully that's ok, but happy to revisit if that still feels too brittle!

File: app/components/blacklight/date_facet_hierarchy_component.rb
Comment (jmartin-sul) [reply app/components/blacklight/date_facet_hierarchy_component.rb:37]: i may have gone too far in the other direction, but i added what should hopefully be a thorough explanation of expected input and output structures 😅

File: app/components/blacklight/date_facet_hierarchy_component.rb
Comment (jmartin-sul) [reply app/components/blacklight/date_facet_hierarchy_component.rb:49]: added some inline comments that try to explain what the conditionals mean in the larger context (of both the current loop iteration and the prior loop iterations), as well as a couple example strings that would meet the condition for skipping to the next loop iteration.

File: spec/features/blacklight_customizations/hierarchical_date_facet_spec.rb
Comment (jmartin-sul) [inline spec/features/blacklight_customizations/hierarchical_date_facet_spec.rb:16]: try to keep the test from breaking when unrelated fixtures are added, while confirming that simply expanding a tree item doesn't facet on that node immediately.

File: spec/features/blacklight_customizations/hierarchical_date_facet_spec.rb
Comment (jmartin-sul) [inline spec/features/blacklight_customizations/hierarchical_date_facet_spec.rb:19]: hopefully we won't regularly be adding or removing fixtures that fall in the 1990-1999 time range, but happy to revisit if thus hardcoded count still feels too brittle.

File: app/components/blacklight/date_facet_hierarchy_component.rb
Comment (jmartin-sul) [inline app/components/blacklight/date_facet_hierarchy_component.rb:31]: happy to edit down if this is overkill 😅

**Review by jmartin-sul — COMMENTED** [review]
@marlo-longley i think this is ready for another look, thanks for the feedback!

i pushed the changes as two new commits, both for easier review, and for easier undo if more iteration is needed.  happy to squash before merge if approved and minimizing the number of new commits is desired.  (but the commits also seem like logically separate chunks of development, so i'm also fine leaving as is)

File: app/components/blacklight/date_facet_hierarchy_component.rb
Comment (marlo-longley) [reply app/components/blacklight/date_facet_hierarchy_component.rb:37]: Ty!!

File: spec/features/blacklight_customizations/hierarchical_date_facet_spec.rb
Comment (marlo-longley) [reply spec/features/blacklight_customizations/hierarchical_date_facet_spec.rb:19]: This seems fine.

**Review by marlo-longley — APPROVED** [review]
This looks good to me.

**Pull request merged**
